doc: fix multiple Doxygen and breathe problems

gmarull opened the pull request:
# Introduction
This PR is an attempt to fix multiple documentation problems we have when using Doxygen + Breathe.

## Summary of the changes
 - Disabled `INLINE_SIMPLE_STRUCTS` Doxygen option. Some struct pages were missing in the Doxygen output (e.g. check https://docs.zephyrproject.org/apidoc/latest/group__bt__gatt__client.html)
 - Removed a duplicate definition in `ieee802154` headers. While valid C code, as of today Sphinx C domain/breathe do not support duplicate definitions. In this case I think the duplicate was likely unintentional.
 - Doxygen (and subsequently Sphinx/Breathe) got confused by the `ATOMIC_DEFINE` inside a struct (see https://docs.zephyrproject.org/apidoc/latest/structbt__gatt__subscribe__params.html). Refactored this part.
 - Introduced a new Sphinx extensions to handle warnings natively, so without the need of post-processing the output.

## Remaining issues

The remaining issues should be fixed by https://github.com/sphinx-doc/sphinx/pull/8313. TLDR; as of today you can't have something like:

```c
 /* demo considered a duplicate because the name is used by both struct and function */
struct demo {
...
};

void demo(...);

/* id considered duplicate */
union {
	uint16_t id;
	struct {
		uint16_t id;
	} vnd;
};
```

After Sphinx fixes the problem above we should be able to (hopefully) get rid of the current "known issues".

## `zephyr.warnings_filter` extension

Instead of post-processing Sphinx output, I have created a small extension that natively filters out Sphinx warnings. This Sphinx plugin can be used to filter out warnings that are known to be false positives. The warnings are filtered out based on a set of regular expressions given via a configuration file, the same principle as before.

At the same time I have simplified current regular expressions and moved them to `doc/known-warnings.txt` with the idea to make docs more self-contained.

gmarull committed:
net: ieee802154: remove duplicate definition

`IEEE802154_MAX_ADDR_LENGTH` was defined in 2 headers, keep one only.

Signed-off-by: Gerard Marull-Paretas <gerard.marull@nordicsemi.no>

gmarull committed:
doc: disable INLINE_SIMPLE_STRUCTS Doxygen option

This option prevents some structures to have missing pages on the
Doxygen HTML output.

Signed-off-by: Gerard Marull-Paretas <gerard.marull@nordicsemi.no>

gmarull committed:
doc: predefine ATOMIC_DEFINE

Predefine ATOMIC_DEFINE in the Doxyfile so that documentation output is
generated correctly. In order to simplify the predefinition
ATOMIC_BITMAP_SIZE has been introduced.

Signed-off-by: Gerard Marull-Paretas <gerard.marull@nordicsemi.no>

jhedberg commented on include/bluetooth/gatt.h:
```diff
@@ -1494,7 +1494,7 @@ struct bt_gatt_subscribe_params {
 	uint16_t value;
 	/** Subscription flags */
-	ATOMIC_DEFINE(flags, BT_GATT_SUBSCRIBE_NUM_FLAGS);
+	atomic_t flags[ATOMIC_BITS_TO_SZ(BT_GATT_SUBSCRIBE_NUM_FLAGS)];
```
Not really happy with the inconsistency that this creates (not to mention it's less readable). There are many other places in Zephyr that use ATOMIC_DEFINE() inside structs. It just seems this is the only struct member with an associated doxygen documentation string. Would it be possible to fix this some other way, e.g. have doxygen/breathe look at the code only after pre-processor expansion?

gmarull replied:
Thanks for the feedback @jhedberg.

I've found other possible solutions for this problem:

1. Use `EXPAND_AS_DEFINED` (ref. https://www.doxygen.nl/manual/config.html#cfg_expand_as_defined). This option requires `EXPAND_ONLY_PREDEF`, though. **After some tests I haven't been able to make it work...**
2. Add the `ATOMIC_DEFINE` macro in `PREDEFINED`, i.e. `ATOMIC_DEFINE(x, y)=atomic_t x[ATOMIC_BITS_TO_SZ(y)]`. This option requires to duplicate the actual macro definition in order to obtain correct Doxygen output. Result:

![image](https://user-images.githubusercontent.com/25011557/113491208-529b9080-94cf-11eb-9d0a-723354df373b.png)

For reference, the current output (incorrect):
![image](https://user-images.githubusercontent.com/25011557/113491179-21bb5b80-94cf-11eb-95e0-c00d5cd8cbc2.png)

Other usages found in `include` that are not rendered because field is hidden by a non-defined Kconfig option.

- include/net/ethernet.h:369:     ATOMIC_DEFINE(interfaces, NET_VLAN_MAX_COUNT);

Other usages found in `include` that do not render well in Doxygen but cause no warnings on Sphinx (maybe API is not exposed in the Sphinx side...?)

- include/net/net_if.h:435:       ATOMIC_DEFINE(flags, NET_IF_NUM_FLAGS);
- include/bluetooth/mesh/cdb.h:40:        ATOMIC_DEFINE(flags, BT_MESH_CDB_NODE_FLAG_COUNT);
- include/bluetooth/mesh/cdb.h:75:        ATOMIC_DEFINE(flags, BT_MESH_CDB_FLAG_COUNT);
- include/bluetooth/l2cap.h:102:  ATOMIC_DEFINE(status, BT_L2CAP_NUM_STATUS);

jhedberg replied:
The output of option 2 above looks quite good to me, in fact I'd say it's better than an explicit integer for the array size since what we're really interested in is how many bits the field can hold.

FWIW, I think the `ATOMIC_DEFINE()` macro itself might be a bit badly named since it's strictly about bitfields and not using atomic_t as a normal integer value. Perhaps `ATOMIC_BITFIELD_DEFINE()` might have been better. Anyway, that's a completely different discussion that doesn't have much to do with this PR.

gmarull replied:
we may use `ATOMIC_BITMAP_SIZE()` instead of `ATOMIC_BITS_TO_SZ()`

jhedberg replied:
Yeah, that sounds good to me

jhedberg replied:
Btw, regarding the naming, Linux has a `DECLARE_BITMAP()` which is essentially exactly the same as our `ATOMIC_DEFINE()`

gmarull committed:
doc: extensions: add warnings_filter

warnings_filter is an extension that allows to filter out Sphinx
warnings.

Signed-off-by: Gerard Marull-Paretas <gerard.marull@nordicsemi.no>

gmarull committed:
doc: use warnings_filter

Replace current filter setup with the warnings_filter Sphinx extension.
Note that current regexes have been simplified to make them more
readable while keeping the most relevant information.

Sphinx warnings are now treated as errors (-W).

Signed-off-by: Gerard Marull-Paretas <gerard.marull@nordicsemi.no>

jukkar reviewed (approved):
LGTM

carlescufi commented on include/sys/atomic.h:
```diff
@@ -68,6 +68,14 @@ typedef void *atomic_ptr_t;
+ * represent at bitmap with @a num_bits.
+ *
+ * @param num_bits Number of bits.
+ */
```
@hidden Please remember to update doxyfile.in if you change this macro

carlescufi commented on include/sys/atomic.h:
```diff
@@ -68,6 +68,14 @@ typedef void *atomic_ptr_t;
+ *
+ * @param num_bits Number of bits.
+ */
+#define ATOMIC_BITMAP_SIZE(num_bits) (1 + ((num_bits) - 1) / ATOMIC_BITS)
```
Merge this with the commit that modifies the doxyfile

gmarull replied:
done

carlescufi commented:
API meeting 2021.04.06

- Remove the commit `net: net_ip: hide some external and forward declarations from Doxygen`
- Otherwise agreed to the changes

marc-hb commented:
While testing #33448 I noticed that `stderr` output from Sphinx and (more commonly) from the very many subprocesses it forks escaped the defunct `filter-known-issues.py` and tended to get drowned in the rest of the output and missed. @gmarull I'm curious whether what is the current status of `stderr` after this PR, unchanged? Just wondering whether you're aware of that since you're on top of everything documentation-related now.

gmarull commented:
> While testing #33448 I noticed that `stderr` output from Sphinx (and more commonly) from the very many subprocesses it forks) escaped the defunct `filter-known-issues.py` and tended to get drowned in the rest of the output and missed. @gmarull I'm curious whether what is the current status of `stderr` after this PR, unchanged? Just wondering whether you're aware of that since you're on top of everything documentation-related now.

Thir PR introduced a way of filtering warnings that does not deal anymore with the output of Sphinx. What it does is to directly intercept Sphinx warning logs and filters them out (or converts them to `INFO` in non silent mode). So the `stderr` will be what Sphinx originally outputs there.
